File: July-LeetCoding-Challenge/24-All-Paths-From-Source-to-Target/All-Paths-From-Source-to-Target.cs
public class Solution {
    public IList<IList<int>> AllPathsSourceTarget(int[][] graph) {
        var res = new List<IList<int>>();
        var path = new List<int>();
        path.Add(0);
        Dfs(graph, 0, res, path);        
        return res;
    }
    
    public void Dfs(int[][] graph, int node, List<IList<int>> res, List<int> path){
        if(node == graph.Length-1) {
            res.Add(new List<int>(path));
            return;
        }
        foreach(var n in graph[node]){
            path.Add(n);
            Dfs(graph, n, res, path);
            path.RemoveAt(path.Count()-1);
        }
    }
}

File: July-LeetCoding-Challenge/24-All-Paths-From-Source-to-Target/All-Paths-From-Source-to-Target.kt
package depth_first_search

import java.util.*
import kotlin.collections.ArrayList

class AllPathsFromSourcetoTargetKotlin797 {
    fun allPathsSourceTarget(graph: Array<IntArray>): List<List<Int>> {
        val result: MutableList<MutableList<Int>> = LinkedList()
        val subsets: MutableList<Int> = LinkedList()

        dfs(result, subsets, graph, 0)
        return result
    }

    private fun dfs(
        result: MutableList<MutableList<Int>>,
        subsets: MutableList<Int>,
        graph: Array<IntArray>,
        index: Int
    ) {
        subsets.add(index)
        if (graph[index].isEmpty()) {
            result.add(ArrayList(subsets))
        } else {
            graph[index].forEach {
                dfs(result, subsets, graph, it)
                subsets.removeAt(subsets.size - 1)
            }
        }
    }
}

fun main() {
    val solution = AllPathsFromSourcetoTargetKotlin797()
    val result = solution.allPathsSourceTarget(
        arrayOf(
            intArrayOf(1, 2),
            intArrayOf(3),
            intArrayOf(3),
            intArrayOf()
        )
    )
    println(result)
}